File: sdks/csharp/Http.cs
// Copyright 2026 J3nna Technologies, LLC
// SPDX-License-Identifier: Apache-2.0
//
// Minimal JSON-over-HTTP helpers shared by Enroll, Discovery, and Mcp. Returns parsed JsonElement trees so
// the SDK can read the loosely-typed console/gossip/JSON-RPC responses without bespoke DTOs.

using System.Text;
using System.Text.Json;

namespace J3nnaMesh;

internal static class Http
{
    private static readonly HttpClient Client = new();

    public static JsonElement Get(string url, TimeSpan timeout)
    {
        using var cts = new CancellationTokenSource(timeout);
        using var resp = Client.GetAsync(url, cts.Token).GetAwaiter().GetResult();
        resp.EnsureSuccessStatusCode();
        var body = resp.Content.ReadAsStringAsync(cts.Token).GetAwaiter().GetResult();
        return JsonDocument.Parse(body).RootElement.Clone();
    }

    public static JsonElement PostJson(string url, string json, TimeSpan timeout)
    {
        using var cts = new CancellationTokenSource(timeout);
        using var content = new StringContent(json, Encoding.UTF8, "application/json");
        using var resp = Client.PostAsync(url, content, cts.Token).GetAwaiter().GetResult();
        resp.EnsureSuccessStatusCode();
        var body = resp.Content.ReadAsStringAsync(cts.Token).GetAwaiter().GetResult();
        return JsonDocument.Parse(body).RootElement.Clone();
    }
}

File: sdks/dart/lib/src/http.dart
// Copyright 2026 J3nna Technologies, LLC
// SPDX-License-Identifier: Apache-2.0
//
// Minimal JSON-over-HTTP helpers on dart:io HttpClient, shared by enroll/discovery/mcp.

import 'dart:async';
import 'dart:convert';
import 'dart:io';

final HttpClient _client = HttpClient();

Future<dynamic> httpGetJson(String url, {Duration timeout = const Duration(seconds: 10)}) async {
  final req = await _client.getUrl(Uri.parse(url));
  final resp = await req.close().timeout(timeout);
  final body = await resp.transform(utf8.decoder).join();
  return jsonDecode(body);
}

Future<dynamic> httpPostJson(String url, Object obj,
    {Duration timeout = const Duration(seconds: 10)}) async {
  final req = await _client.postUrl(Uri.parse(url));
  req.headers.set(HttpHeaders.contentTypeHeader, 'application/json');
  req.add(utf8.encode(jsonEncode(obj)));
  final resp = await req.close().timeout(timeout);
  final body = await resp.transform(utf8.decoder).join();
  return jsonDecode(body);
}
